core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/ForwardChainingSchemaCachingRDFSInferencer.java

```diff
@@ -0,0 +1,318 @@
+
+    // The schema, or null
+    Repository schema;
+
+    private static final Random random = new Random();
+
+    private StampedLock readWriteLock = new StampedLock();
+
+    // If false, the inferencer will skip some RDFS rules.
+    boolean useAllRdfsRules = true;
+
+    // Tbox cache
+    Collection<Resource> properties = new HashSet<>();
+    Collection<Resource> types = new HashSet<>();
+    Collection<Statement> subClassOfStatements = new ArrayList<>();
+    Collection<Statement> subPropertyOfStatements = new ArrayList<>();
+    Collection<Statement> rangeStatements = new ArrayList<>();
+    Collection<Statement> domainStatements = new ArrayList<>();
+
+    // Forward chained Tbox cache as lookup tables
+    Map<Resource, Set<Resource>> calculatedTypes = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedProperties = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedRange = new HashMap<>();
+    Map<Resource, Set<Resource>> calculatedDomain = new HashMap<>();
+
+    // The inferencer has been instantiated from another inferencer and shares it's schema with that one
+    private boolean sharedSchema;
+
+    void clearInferenceTables() {
+        properties = new HashSet<>();
+        types = new HashSet<>();
+        subClassOfStatements = new ArrayList<>();
+        subPropertyOfStatements = new ArrayList<>();
+        rangeStatements = new ArrayList<>();
+        domainStatements = new ArrayList<>();
+
+        calculatedTypes = new HashMap<>();
+        calculatedProperties = new HashMap<>();
+        calculatedRange = new HashMap<>();
+        calculatedDomain = new HashMap<>();
+    }
+
+    void readLock(ForwardChainingSchemaCachingRDFSInferencerConnection connection) {
+
+//        if (numberOfThreadsWaitingForWriteLock.get() > 0) {
+//            System.err.println("starve reads");
+//        }
+
+        while (numberOfThreadsWaitingForWriteLock.get() > 0) {
+            Thread.yield();
+        }
+
+        if (connection.lockStamp != 0) {
+            throw new SailException("Connection already has a lock! Might be a begin() without a commit() or rollback() for previous transaction().");
+        }
+        connection.lockStamp = readWriteLock.readLock();
+//        System.err.println("readLock: " + connection.lockStamp);
+
+    }
+
+    void releaseLock(ForwardChainingSchemaCachingRDFSInferencerConnection connection) {
+        if(connection.lockStamp == 0){
+            throw new SailException("Expected connection to have lock. Might be a commit() without a begin().");
+        }
+        readWriteLock.unlock(connection.lockStamp);
+//        System.err.println("Released lock: " + connection.lockStamp);
+
+        connection.lockStamp = 0;
+    }
+
+
+    private AtomicInteger numberOfThreadsWaitingForWriteLock = new AtomicInteger(0);
+
+    void upgradeLock(ForwardChainingSchemaCachingRDFSInferencerConnection connection) {
```

Comment:
I am still unsure in general whether there needs to be an ad-hoc locking scheme implemented here. Ie, it isn't clear why you can't have a single ReentrantReadWriteLock and rely on it rather than creating a novel unsynchronised scheme here from first principles.